src/HEVM.hs

```diff
@@ -1,446 +1,365 @@
-          VMSuccess (SymbolicBuffer msg) -> (msg, sFalse)
-          VMFailure (Revert _) -> ([], sTrue)
-          _ -> ([], sFalse)
-        expected = maybe [] (toSymBytes . symExp ctx) returns
-
-    in (preCondPass .=>
-        (postCond' .&& storageConstraints .&& sNot reverted .&& (actual .== expected))) .&&
-       (preCondFail .=> reverted) -- TODO shall we check for postCond' .&& storageConstraints here?
-
--- | Locate the variables refered to in the act-spec in the vm
-mkVmContext :: SolcJson -> Behaviour -> Map Id Addr -> VM -> VM -> (Ctx, VMResult)
-mkVmContext solcjson b@(Behaviour method c1 (Interface _ decls) _ _ _ updates _) contractMap pre post =
-  let args = fromList $ locateCalldata b decls (fst $ view (state . calldata) pre) <$> decls
-      env' = makeVmEnv b pre
-      -- we should always have a result after executing the vm fully.
-      Just res = view result post
-      -- add storage entries to the 'store' map as we go along
-      ctx = foldl
-       (\ctx'@(Ctx c m a s e) update' ->
-         let
-           (name, entry) = locateStorage ctx' solcjson contractMap method (pre, post) update'
-         in Ctx c m a (Map.insert name entry s) e)
-       (Ctx c1 method args mempty env') updates
-  in (ctx, res)
-
-
-makeVmEnv :: Behaviour -> VM -> Map Id SActType
-makeVmEnv (Behaviour method c1 _ _ _ _ _ _) vm =
-  fromList
-    [ Caller    |- SymInteger (sFromIntegral $ saddressWord160 (view (state . caller) vm))
-    , Callvalue |- let S _ w = view (state . callvalue) vm
-                   in SymInteger (sFromIntegral w)
-    , Calldepth |- SymInteger (num $ length (view frames vm))
-     -- the following environment variables are always concrete in hevm right now
-    , Origin    |- SymInteger (num $ addressWord160 (view (tx . origin) vm))
-    , Difficulty |- SymInteger (num $ view (block . difficulty) vm)
-    , Chainid |- SymInteger (num $ view (env . chainId) vm)
-    , Timestamp |- let S _ w = view (block . timestamp) vm
-                   in SymInteger (sFromIntegral w)
-    , This |- SymInteger (num $ addressWord160 (view (state . contract) vm))
-    , Nonce |- let
-                 maybeContract = view (env . contracts . at (view (state . contract) vm)) vm
-               in SymInteger $ maybe 1 (num . view nonce) maybeContract
-
-      -- and this one does not even give a reasonable result
---    , Blockhash |- error "blockhash not available in hevm right now"
-    ]
+    mkArg :: Decl -> CalldataFragment
+    mkArg (Decl typ x)  = symAbiArg (T.pack x) typ
+    makeSig = T.pack $ makeIface iface
+    calldatas = fmap mkArg decls
+    (cdBuf, props) = combineFragments calldatas (Types.ConcreteBuf "")
+    withSelector = writeSelector cdBuf makeSig
+    sizeConstraints
+      = (bufLength withSelector Types..>= cdLen calldatas)
+        Types..&& (bufLength withSelector Types..< (Types.Lit (2 ^ (64 :: Integer))))
+  in (withSelector, sizeConstraints : props)
+
+makeCtrCalldata :: Interface -> Calldata
```

Comment:
whats the difference between this and `makeCalldata`?

Reply:
The `makeCtrCalldata` does not have a function sector and it is intended to create the calldata for constructors. It is currently not used since we are not handling constructors yet. I do not remember the exact calling conventions for constructors but it wrote this after some conversations in the Berlin office. If we end up keeping this some things can be written more generically (e.g., `combineFragments`) for avoid code duplication between these two functions.